hydra-explorer/src/Hydra/Explorer.hs

```diff
@@ -68,41 +79,68 @@ logMiddleware tracer app' req sendResponse = do
 
 httpApp :: Tracer IO APIServerLog -> GetHeads -> Application
 httpApp tracer getHeads =
-  logMiddleware tracer $ serve api $ server getHeads
+  logMiddleware tracer $ serve explorerAPI $ server getHeads
 
-observerHandler :: TVar IO ExplorerState -> [HeadObservation] -> IO ()
-observerHandler explorerState observations = do
+observerHandler ::
+  TVar IO ExplorerState ->
+  PersistenceIncremental (OnChainTx Tx) IO ->
+  [HeadObservation] ->
+  IO ()
+observerHandler explorerState PersistenceIncremental{append} observations = do
+  let onChainTxs = mapMaybe convertObservation observations
+  forM_ onChainTxs append
   atomically $
-    modifyTVar' explorerState $
-      aggregateHeadObservations observations
+    modifyTVar' explorerState $ \currentState ->
+      foldl' aggregateOnChainTx currentState onChainTxs
 
 readModelGetHeadIds :: TVar IO ExplorerState -> GetHeads
 readModelGetHeadIds = readTVarIO
 
 main :: IO ()
 main = do
-  withTracer (Verbose "hydra-explorer") $ \tracer -> do
-    explorerState <- newTVarIO (mempty :: ExplorerState)
+  withTracer (Verbose "hydra-explorer") $ \(tracer :: Tracer IO (HydraLog Tx ())) -> do
+    opts <- execParser hydraExplorerOptions
+    let Options
+          { networkId
+          , host = Host{hostname, port}
+          , nodeSocket
+          , startChainFrom
+          , persistenceDir
+          } = opts
+    persistence <- createPersistenceIncremental $ persistenceDir <> "/explorer-state"
+    explorerState <- do
+      let nodeTracer = contramap Node tracer
+      events <- loadAll persistence
+      traceWith nodeTracer LoadedState{numberOfEvents = fromIntegral $ length events}
+      let initialState = mempty
+          recoveredSt = foldl' aggregateOnChainTx initialState events
+      newTVarIO recoveredSt
```

Comment:
Must: include chain information in persisted data to avoid inconsistencies by design

This `explorer-state` seems not to include any information about which point on the chain was last seen. Hence, we can easily "miss heads" if we start the explorer, stop it and then start it again after a few minutes (e.g. when doing maintenance)

This must be avoided! Having an explorer synchronizing everything from a fixed point (`--start-chain-from` or even genesis) is strictly more correct than this half-way persisting. Also, it is not a big problem if starting the explorer takes a few minutes before it is fully live (can be controlled by the `--start-chain-from` option).